heron/proto/ckptmgr.proto

```diff
@@ -0,0 +1,200 @@
+}
+
+// message stored in the state manager by the tmaster
+message StatefulConsistentCheckpoints {
+  // An ordered list of the globally consistent checkpoints
+  // that have been snapshotted and can be recovered from.
+  // 0th element is the most recent one, 1st is the next recent
+  // and so on
+  repeated StatefulConsistentCheckpoint consistent_checkpoints = 1;
+}
+
+// tmaster -> stmgr messages
+
+// Message sent to stmgrs by the tmaster to initiate checkpointing
+message StartStatefulCheckpoint {
+  required string checkpoint_id = 1;
+}
+
+// Message sent by tmaster to stmgr asking them to reset their instances
+// to this checkpoint
+message RestoreTopologyStateRequest {
+  required string checkpoint_id = 1;
+  // Every Attempt made by Tmaster to restore to a globally consistent
+  // checkpoint is tagged with a unique id. The stmgrs include this in their
+  // RestoreTopologyStateResponse below. This way tmaster could
+  // try restoring to the same checkpoint_id multiple times and still
+  // keep track of for which attempt the restore was in the responses
+  required int64 restore_txid = 2; 
+}
+
+// Message that stmgr sends to tmaster after it restores
+// all of its local instances to a checkpoint_id
+message RestoreTopologyStateResponse {
+  required heron.proto.system.Status status = 1;
+  required string checkpoint_id = 2;
+  required int64 restore_txid = 3; 
+}
+
+// Message sent by stmgr to tmaster asking it to reset the topology
+// to some valid checkpoint. This is sent either if stmgr dies
+// and comes back up or if an instance dies.
+message ResetTopologyState {
+  optional string dead_stmgr = 1;
+  optional int32 dead_taskid = 2;
+}
+
+// Message sent by stmgr to tmaster informing it about
+// the fact that we stored a checkpoint belonging
+// to the instance
+message InstanceStateStored {
+  required string checkpoint_id = 1;
+  required heron.proto.system.Instance instance = 2;
+}
+
+// Message sent by tmaster to stmgr to start processing
+// For exactly once semantics, all the topology components
+// should be rolled back to a consistent state before they can
+// start processing.
+message StartStmgrStatefulProcessing {
+  required string checkpoint_id = 1;
+}
+
+// tmaster -> ckptmgr messages
+
+// This is the message that a tmaster sends
+// when it wants to establish the connection
+// with checkpoint manager
+message RegisterTMasterRequest{
+  required string topology_name = 1;
+  required string topology_id = 2;
+}
+
+// This is the message that checkpoint manager
+// sends when it receives the register request
+// from tmaster
+message RegisterTMasterResponse {
+  required heron.proto.system.Status status = 1;
+}
+
+// Message sent by tmaster to ckptmgr to cleanup
+// old checkpoint state.
+message CleanStatefulCheckpointRequest {
+  // Any checkpoints older than this can be cleaned
+  optional string oldest_checkpoint_preserved = 1;
+  optional bool clean_all_checkpoints = 2 [default = false];
```

Comment:
It's  listed in the version 2 spec:
https://developers.google.com/protocol-buffers/docs/reference/proto2-spec

What's the minimum version that supports it? We should consider upgrading, since this was included in 2.x.